python/ray/ml/preprocessor.py

```diff
@@ -1,55 +1,104 @@
 import abc
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import pandas as pd
 
 from ray.data import Dataset
 from ray.ml.predictor import DataBatchType
 
 
+class PreprocessorAlreadyFittedException(Exception):
+    pass
+
+
+class PreprocessorNotFittedException(Exception):
+    pass
+
+
 class Preprocessor(abc.ABC):
-    """Preprocessor interface"""
+    """Preprocessor interface for transforming Datasets."""
 
-    def fit_transform(self, ds: Dataset) -> Dataset:
-        """Fit this preprocessor to the given dataset.
+    # Preprocessors that do not need to be fitted must override this.
+    _is_fittable = True
 
-        Returns the fitted dataset.
+    def fit(self, dataset: Dataset) -> "Preprocessor":
+        """Fit this Preprocessor to the Dataset.
 
-        Args:
-            ds (Dataset): Input dataset.
+        Fitted state attributes will be directly set in the Preprocessor.
 
         Returns:
-            Dataset: Transformed dataset.
-
+            Preprocessor: The fitted Preprocessor with state attributes.
         """
-        raise NotImplementedError
+        if self.check_is_fitted():
+            raise PreprocessorAlreadyFittedException(
+                "`fit` cannot be called multiple times. "
+                "Create a new Preprocessor to fit a new Dataset."
+            )
 
-    def transform(self, ds: Dataset) -> Dataset:
-        """Transforms the given dataset.
+        return self._fit(dataset)
 
-        The preprocessor must be fitted first.
+    def _fit(self, dataset: Dataset) -> "Preprocessor":
```

Comment:
Just a nit, but you may want to consider marking these as abstractmethods: https://docs.python.org/3.8/library/abc.html#abc.abstractmethod.

This would require subclasses to override these methods.

Reply:
I thought about this but then this would _require_ subclasses to override `_fit`. I wasn't sure if there would be a case where they'd need to override `fit` entirely instead. This is probably more applicable to the `_transform_pandas` method though.